### ISSUE TITLE
[Bug] Segfault on CPU BackwardBinaryReduce with -O0 compiler switch

### ISSUE THREAD

**Issue by BarclayII** [dmlc/dgl#1056]
## 🐛 Bug

If compiler optimization is turned off ('-O0'), DGL would segfault when computing gradients of builtin kernels.

## To Reproduce

1. In `CMakeLists.txt`, replace the two `-O2`'s with `-g -O0`.
2. Build and install normally with CMake and `make`
3. Run the following and observe crash.

```python
import dgl
import numpy as np
import scipy.sparse as ssp
import torch
import torch.nn as nn
import tqdm

device = torch.device('cpu')
g = dgl.DGLGraph(ssp.random(2000, 2000, 0.02))
m = nn.Linear(1000, 2000).to(device)

for _ in tqdm.trange(10000):
    x = torch.randn(2000, 1000).to(device)
    g.ndata['x'] = x
    g.ndata['w'] = m(x)
    g.update_all(dgl.function.copy_u('w', 'm'), dgl.function.sum('m', 'y'))
    loss = g.ndata['y'].sum()
    loss.backward()
```

## Expected behavior

Shouldn't crash

## Environment

 - DGL Version (e.g., 1.0): master
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3):
 - OS (e.g., Linux): Ubuntu 16.04 and CentOS 7
 - How you installed DGL (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): see above
 - Python version: 3.6
 - CUDA/cuDNN version (if applicable):
 - GPU models and configuration (e.g. V100):
 - Any other relevant information:

## Additional context

GPU kernels do not crash.

Both OpenMP and non-OpenMP builds would crash.

**Would not crash with -O2 option.**

Relevant stacktrace:
```
#0  0x00007fffa11cb3bb in dgl::kernel::cpu::BackwardFunctorsTempl<int, float, dgl::kernel::SelectDst, dgl::kernel::SelectNone, dgl::ker
nel::BinaryUseLhs<float>, dgl::kernel::ReduceSum<1, float> >::Read (addr=0x7fff29e8e000)
    at /mnt/b/gq/dgl/src/kernel/cpu/./backward_binary_reduce_impl.h:185
#1  0x00007fffa11a0102 in dgl::kernel::cpu::BackwardBinaryReduce<0, int, float, dgl::kernel::cpu::BackwardFunctorsTempl<int, float, dgl
::kernel::SelectDst, dgl::kernel::SelectNone, dgl::kernel::BinaryUseLhs<float>, dgl::kernel::ReduceSum<1, float> > >::ApplyEdge (
    src=0, dst=27, eid=0, gdata=0x7fffb0568620) at /mnt/b/gq/dgl/src/kernel/cpu/./backward_binary_reduce_impl.h:70
#2  0x00007fffa118e7cd in minigun::advance::CPUAdvance<int, minigun::advance::Config<true, (minigun::advance::FrontierMode)0>, dgl::ker
nel::BackwardGData<int, float>, dgl::kernel::cpu::BackwardBinaryReduce<0, int, float, dgl::kernel::cpu::BackwardFunctorsTempl<int, floa
t, dgl::kernel::SelectDst, dgl::kernel::SelectNone, dgl::kernel::BinaryUseLhs<float>, dgl::kernel::ReduceSum<1, float> > >, minigun::De
faultAllocator<1> > (csr=..., gdata=0x7fffb0568620, input_frontier=..., output_frontier=..., lcl_row_offsets=...,
    alloc=0x7fffa3fd9f7a <minigun::DefaultAllocator<1>::Get()::alloc>)
    at /mnt/b/gq/dgl/third_party/minigun/minigun/minigun/././cpu/advance.h:89
#3  0x00007fffa116d2cb in minigun::advance::DispatchXPU<1, int, minigun::advance::Config<true, (minigun::advance::FrontierMode)0>, dgl:
:kernel::BackwardGData<int, float>, dgl::kernel::cpu::BackwardBinaryReduce<0, int, float, dgl::kernel::cpu::BackwardFunctorsTempl<int,
float, dgl::kernel::SelectDst, dgl::kernel::SelectNone, dgl::kernel::BinaryUseLhs<float>, dgl::kernel::ReduceSum<1, float> > >, minigun
::DefaultAllocator<1> >::Advance (rtcfg=..., csr=..., gdata=0x7fffb0568620, input_frontier=..., output_frontier=0x0,
    alloc=0x7fffa3fd9f7a <minigun::DefaultAllocator<1>::Get()::alloc>)
    at /mnt/b/gq/dgl/third_party/minigun/minigun/minigun/././cpu/advance.h:164
#4  0x00007fffa115d9a7 in minigun::advance::Advance<1, int, minigun::advance::Config<true, (minigun::advance::FrontierMode)0>, dgl::ker
nel::BackwardGData<int, float>, dgl::kernel::cpu::BackwardBinaryReduce<0, int, float, dgl::kernel::cpu::BackwardFunctorsTempl<int, floa
t, dgl::kernel::SelectDst, dgl::kernel::SelectNone, dgl::kernel::BinaryUseLhs<float>, dgl::kernel::ReduceSum<1, float> > >, minigun::De
faultAllocator<1> > (config=..., csr=..., gdata=0x7fffb0568620, input_frontier=..., output_frontier=0x0,
    alloc=0x7fffa3fd9f7a <minigun::DefaultAllocator<1>::Get()::alloc>)
    at /mnt/b/gq/dgl/third_party/minigun/minigun/minigun/./advance.h:113
#5  0x00007fffa114cfb0 in dgl::kernel::CallBackwardBinaryReduce<1, 0, int, float, dgl::kernel::SelectSrc, dgl::kernel::SelectNone, dgl:
:kernel::BinaryUseLhs<float>, dgl::kernel::ReduceSum<1, float> > (rtcfg=..., graph=..., gdata=0x7fffb0568620)
    at /mnt/b/gq/dgl/src/kernel/cpu/./backward_binary_reduce_impl.h:243
#6  0x00007fffa11f0859 in dgl::kernel::BackwardBinaryReduceImpl<1> (reducer=..., op=..., graph=..., lhs=dgl::kernel::binary_op::kSrc,
    rhs=dgl::kernel::binary_op::kNone, lhs_mapping=..., rhs_mapping=..., out_mapping=..., lhs_data=..., rhs_data=..., out_data=...,
    grad_out_data=..., grad_lhs_data=..., grad_rhs_data=...) at /mnt/b/gq/dgl/src/kernel/cpu/../binary_reduce_impl.h:205
#7  0x00007fffa0b5d090 in dgl::kernel::BackwardCopyReduce (reducer=..., graph=..., target=dgl::kernel::binary_op::kSrc,
    in_mapping=..., out_mapping=..., in_data=..., out_data=..., grad_out_data=..., grad_in_data=...)
    at /mnt/b/gq/dgl/src/kernel/binary_reduce.cc:590
#8  0x00007fffa0b5dd1d in dgl::kernel::<lambda(dgl::runtime::DGLArgs, dgl::runtime::DGLRetValue*)>::operator()(dgl::runtime::DGLArgs, d
gl::runtime::DGLRetValue *) const (__closure=0x555555ec95e0, args=..., rv=0x7fffb0569320)
    at /mnt/b/gq/dgl/src/kernel/binary_reduce.cc:609
[snip since the rest are mostly on Python side...]
```

**Comment by VoVAllen** [comment]
Seems the problem with openmp with multi threading

**Comment by VoVAllen** [comment]
![image](https://user-images.githubusercontent.com/8686776/69791983-9bb74280-1200-11ea-9a5c-91f4700c828b.png)

**Comment by BarclayII** [comment]
> Seems the problem is due to openmp with multi threading

The problem persists when OpenMP is turned off (i.e. CMake option `-DUSE_OPENMP=OFF`)

**Comment by VoVAllen** [comment]
I tried `-DUSE_OPENMP=OFF` and recompile it. It works fine afterward.

**Comment by BarclayII** [comment]
> I tried `-DUSE_OPENMP=OFF` and recompile it. It works fine afterward.

Strange... because my stacktrace is generated without OpenMP (hence no "libgomp" in the output)

**Comment by BarclayII** [comment]
Does not segfault when compiled with GCC 9+.

Maybe this problem is related to the version of the compiler...

**Comment by BarclayII** [comment]
Here's my guess based on some gdb'ing.

In `src/kernel/binary_reduce.cc::BackwardCopyReduce()`:

https://github.com/dmlc/dgl/blob/10b104dd35cb2c03fc07feb24d928721eb077d07/src/kernel/binary_reduce.cc#L590-L595

Since we are actually computing gradient of a copy-reduce operator, we reuse the backward operator for binary message functions, filling the non-existent RHS array with an empty one.

Then after multiple redirections, we get to the `ApplyEdge()` function in `src/kernel/cpu/backward_binary_reduce_impl.h` with `Mode == binary_op::kGradLhs`:

https://github.com/dmlc/dgl/blob/10b104dd35cb2c03fc07feb24d928721eb077d07/src/kernel/cpu/backward_binary_reduce_impl.h#L143-L149

where we read from the empty RHS array, causing an invalid memory access under `-O0`.

And because the following `Functors::BackwardOpLhs` in fact dispatches to `BinaryUseLhs::BackwardLhs` which always returns 1 independent of the RHS array, GCC with `-O2` skips both reads, thus getting around the invalid memory access.

Does this make sense?  @jermainewang @yzh119 @classicsong I'm not as familiar as you guys in terms of the kernels.

**Comment by classicsong** [comment]
Yes. It seems a problem here if rhs_base is None. And -O2 will remove the dead code.

**Comment by jermainewang** [comment]
Confirm this is the problem. My bad. I was aware of this issue and should leave a note there. There might be bugs in the forward function too. I can think of two ways to fix this.
* Add a guarding if before the `Functors::Read`. However, we need to know whether the `BackwardOp` uses the lhs or rhs.
* Fuse the Read op to the `BackwardOp`. Essentially passing the pointers to the functor.

Both fixes require some efforts though.
